src/websockethandler.cpp

```diff
@@ -0,0 +1,178 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "websockethandler.h"
+#include "networkrequest.h"
+#include "settingsholder.h"
+#include "urlopener.h"
+#include "mozillavpn.h"
+
+#include <QtCore/QDebug>
+#include <QAbstractSocket>
+#include <QTimer>
+
+constexpr const char* WEBSOCKET_SERVER_URL = "ws://localhost:3000";
+constexpr const int WEBSOCKET_PING_INTERVAL = 30 * 1000;   // 30s
+constexpr const int WEBSOCKET_RETRY_INTERVAL = 30 * 1000;  // 30s
+
+WebSocketHandler::WebSocketHandler() {
+  connect(&m_webSocket, &QWebSocket::connected, this,
+          &WebSocketHandler::onConnected);
+  connect(&m_webSocket, &QWebSocket::disconnected, this,
+          &WebSocketHandler::onClose);
+  connect(&m_webSocket,
+          QOverload<QAbstractSocket::SocketError>::of(&QWebSocket::error), this,
+          &WebSocketHandler::onError);
+  connect(&m_webSocket, &QWebSocket::pong, this, &WebSocketHandler::onPong);
+}
+
+/**
+ * @brief Opens the websocket connection.
+ *
+ * No-op in case the connection is already open.
+ */
+void WebSocketHandler::open() {
+  if (m_webSocket.state() != QAbstractSocket::UnconnectedState &&
+      m_webSocket.state() != QAbstractSocket::ClosingState) {
+    qDebug()
```

Comment:
Would you mind expanding on why? I am relying on the user state, because I need the auth token to connect to the websocket. The `State` enum has many more variants, UserState looks more straight forward for this case IMO.